tooling/nargo_cli/src/cli/test_cmd.rs

```diff
@@ -77,29 +88,65 @@ pub(crate) fn run(args: TestCommand, config: NargoConfig) -> Result<(), CliError
         None => FunctionNameMatch::Anything,
     };
 
-    // Configure a thread pool with a larger stack size to prevent overflowing stack in large programs.
-    // Default is 2MB.
-    let pool = rayon::ThreadPoolBuilder::new().stack_size(4 * 1024 * 1024).build().unwrap();
-    let test_reports: Vec<Vec<(String, TestStatus)>> = pool.install(|| {
-        workspace
-            .into_iter()
-            .par_bridge()
-            .map(|package| {
-                run_tests::<Bn254BlackBoxSolver>(
-                    &workspace_file_manager,
-                    &parsed_files,
-                    package,
-                    pattern,
-                    args.show_output,
-                    args.oracle_resolver.as_deref(),
-                    Some(workspace.root_dir.clone()),
-                    Some(package.name.to_string()),
-                    &args.compile_options,
-                )
-            })
-            .collect::<Result<_, _>>()
-    })?;
-    let test_report: Vec<(String, TestStatus)> = test_reports.into_iter().flatten().collect();
+    let mut tests = Vec::new();
+    for package in workspace.into_iter() {
+        let package_tests = collect_package_tests::<Bn254BlackBoxSolver>(
+            &workspace_file_manager,
+            &parsed_files,
+            package,
+            pattern,
+            args.show_output,
+            args.oracle_resolver.as_deref(),
+            Some(workspace.root_dir.clone()),
+            package.name.to_string(),
+            &args.compile_options,
+        )?;
+        tests.extend(package_tests);
+    }
+
+    let num_tests = tests.len();
+
+    let num_threads = std::thread::available_parallelism().ok().map(Into::into).unwrap_or(1);
+
+    let (sender, receiver) = mpsc::channel();
+    let iter = Mutex::new(tests.into_iter());
```

Comment:
Maybe consider using [crossbeam channels](https://docs.rs/crossbeam/latest/crossbeam/channel/index.html#sharing-channels) which can be shared between multiple consumers, so there would be no need for an explicit mutex around an iterator. You could even start running tests while they are being collected.

Reply:
Maybe this could be done in a separate PR?